### PR TITLE
Implement CreditCode and CreditReward models and repos

### PR DESCRIPTION
**Pull request by leogrim**
@andrewconner @stkem

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+import scala.concurrent.duration.Duration
+import scala.util.Try
+
+case class CreditCode(value: String) {
+  def urlEncoded: String = URLEncoder.encode(value, UTF8)
+}
+
+object CreditCode {
+  def normalize(rawCode: String): CreditCode = CreditCode(rawCode.toLowerCase.trim)
+  implicit def format: Format[CreditCode] = Format(Reads.of[String].map(normalize(_)), Writes[CreditCode](code => JsString(code.value)))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala]: `val` instead of `def`?

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  def columnType(db: DataBaseComponent) = {
+    import db.Driver.simple._
+    MappedColumnType.base[CreditCode, String](_.value, CreditCode.apply)
+  }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala:28]: Why is this here instead of in the repo or in the generic typemappers file?

Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala:28]: It's not in common and it's used in two different repos, DRYing up.

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+sealed abstract class CreditCodeKind(val kind: String)
+object CreditCodeKind {
+  case object Coupon extends CreditCodeKind("coupon")
+  case object OrgReferral extends CreditCodeKind("org_referral")
+
+  def isSingleUse(kind: CreditCodeKind) = kind match {
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala:36]: Maybe include a `numberOfUses: Option[Int]` in the credit code kind itself?

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala]: :+1:

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+  private val all: Set[CreditCodeKind] = Set(Coupon, OrgReferral)
+
+  def apply(kind: String) = all.find(_.kind equalsIgnoreCase kind) match {
+    case Some(validKind) => validKind
+    case None => throw new IllegalArgumentException(s"Unknown CreditCodeKind: $kind")
+  }
+}
+
+case class CreditCodeReferrer(userId: Id[User], organizationId: Option[Id[Organization]], credit: Option[DollarAmount])
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala:49]: A little semantically unclear, but as you said in person it only shows up in this one place.

Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala:49]: This is used to generate a referrer reward if necessary.

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    info.referrer.flatMap(_.credit)
+  ))
+}
+
+case class UnavailableCreditCodeException(info: CreditCodeInfo) extends Exception(s"Credit code ${info.code} is unavailable, already exists with kind ${info.kind.kind}")
+
+@ImplementedBy(classOf[CreditCodeInfoRepoImpl])
+trait CreditCodeInfoRepo extends Repo[CreditCodeInfo] {
+  def getByCode(code: CreditCode, excludeState: Option[State[CreditCodeInfo]] = Some(CreditCodeInfoStates.INACTIVE))(implicit session: RSession): Option[CreditCodeInfo]
+  def add(code: CreditCode, kind: CreditCodeKind, credit: DollarAmount, validityPeriod: Option[Duration], referrer: Option[CreditCodeReferrer])(implicit session: RWSession): Try[CreditCodeInfo]
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala]: I'm wary of including this pseudo-commander logic in the repo itself. I feel like this an awkward way of avoiding `creditCodeInfoRepo.save(CreditCodeInfo(code, kind, credit, validityPeriod, referrer))`, and it's not really the place of a repo to deal with errors (presumably issues with the unique index on the `code`?).

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+@Singleton
+class CreditCodeInfoRepoImpl @Inject() (
+  val db: DataBaseComponent,
+  val clock: Clock)
+    extends DbRepo[CreditCodeInfo] with CreditCodeInfoRepo with Logging {
+
+  import db.Driver.simple._
+
+  implicit val creditCodeTypeMapper = CreditCode.columnType(db)
+  implicit val dollarAmountColumnType = DollarAmount.columnType(db)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala:132]: I feel like these should either be declare right here (if this is the only place they're used) or put in `FortyTwoGenericTypeMappers`. It's weird to define the db stuff in the model.

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    def referrerCredit = column[Option[DollarAmount]]("referrer_credit", O.Nullable)
+    def * = (id.?, createdAt, updatedAt, state, code, kind, credit, status, validityPeriod, referrerUserId, referrerOrganizationId, referrerCredit) <> ((CreditCodeInfo.applyFromDbRow _).tupled, CreditCodeInfo.unapplyToDbRow)
+  }
+
+  def table(tag: Tag) = new CreditCodeInfoTable(tag)
+  initTable()
+
+  override def deleteCache(info: CreditCodeInfo)(implicit session: RSession): Unit = {}
+
+  override def invalidateCache(info: CreditCodeInfo)(implicit session: RSession): Unit = {}
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala:154]: please jesus never implement a cache for these models

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+    updatedAt: DateTime,
+    state: State[CreditReward],
+    userId: Option[Id[User]],
+    organizationId: Option[Id[Organization]],
+    accountId: Id[PaidAccount],
+    credit: DollarAmount,
+    kind: RewardKind,
+    status: String,
+    info: Option[JsValue],
+    unrepeatable: Option[Boolean],
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala]: `repeatable` instead of `unrepeatable`. Easier to reduce the number of negations.

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala (hunk 3 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala:36]: This is tracked by the number of corresponding rewards created.

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala (hunk 8 above)
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala]: :+1:

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+object CreditRewardStates extends States[CreditReward]
+
+case class CreditReward(
+    id: Option[Id[CreditReward]] = None,
+    createdAt: DateTime = currentDateTime,
+    updatedAt: DateTime = currentDateTime,
+    state: State[CreditReward] = CreditRewardStates.ACTIVE,
+    userId: Option[Id[User]],
+    organizationId: Option[Id[Organization]],
+    accountId: Id[PaidAccount],
```
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala:177]: Why do you need this if you have the org id?

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala (hunk 5 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditCode.scala]: After talking about it with you, I'd be happy if this were just 
`def add(model: CreditCodeInfo)(implicit session: RWSession): Try[CreditCodeInfo]`

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala (hunk 8 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala]: Probably need unrepeatable if you want to continue using the `trueOrNull` convention...

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala (hunk 9 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala:177]: referential integrity constraint is more semantically meaningful, and it makes it clearer where the credit is actually going to go

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala:177]: Tough call, though, I know you want to hide `PaidAccount` as much as possible. Here it seems very directly applicable, though.

Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala:177]: @stkem I would like to get rid of it, but I want an unique index on (accountId, kind, unrepeatable), i.e. two rewards of the same unrepeatable kind cannot be applied to the same account (e.g. the reward that you get at org creation).

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala (hunk 8 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala]: That :)

File: kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala (hunk 9 above)
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala:177]: There is a one-to-one mapping of orgId <-> accountId. Why not put the constraint on (orgId, kind, unrepeatable)?

Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/CreditReward.scala:177]: Planning for this to support individual user rewards as well. I could have two different booleans `unrepeatableForOrg` and `unrepeatableForUser` (later on I guess), and maybe renaming `singleUse` to `unrepeatable` for naming consistency). Seems like dealing with an account id is slightly clearer / safer.
